File: Stack/2. Min Stack/solution.scala
class MinStack() {
  import scala.collection.mutable.Stack

  val stack = Stack[(Int, Int)]()

  def push(`val`: Int) {
    val minElement = if(stack.nonEmpty && stack.top._2 < `val`) stack.top._2 else `val`
    val newElement = (`val`, minElement)
    stack.push(newElement)
  }

  def pop() {
    stack.pop
  }

  def top(): Int = {
    if(stack.nonEmpty) stack.top._1
    else throw new NoSuchElementException("Stack is empty")
  }

  def getMin(): Int = {
    if(stack.nonEmpty) stack.top._2
    else throw new NoSuchElementException("Stack is empty")
  }
}

File: Stack/2. Min Stack/solution.rb
class MinStack
    Item = Struct.new(:value, :min, :next)
    def initialize()
      @head = nil
    end


=begin
    :type val: Integer
    :rtype: Void
=end
    def push(val)
      @head = Item.new(value, [value, @head&.min].compact.min, @head)
    end


=begin
    :rtype: Void
=end
    def pop()
      @head = @head.next
    end


=begin
    :rtype: Integer
=end
    def top()
      @head.value
    end


=begin
    :rtype: Integer
=end
    def get_min()
      @head.min
    end
end

# Your MinStack object will be instantiated and called as such:
# obj = MinStack.new()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.get_min()
